Explorer: support index-based fields for `FieldPath` and `Members`

Pixep opened the pull request:
Features
* Support addressing positional members (i.e. tuple fields) using an index.
* Addresses a couple of `TODO`s relative to positional members

Changes:
* Add a new `IndexedValue` mirroring `NamedValue`
* Adds `FieldPath::index() -> int`
* Adds a `Member` variant for `IndexedValue`

Pixep committed:
Explorer: support index-base fields and Members

Add `FieldPath::index() -> std::optional<size_t>`, a `Member`
variant for index-based, and `IndexedValue` mirroring `NamedValue`.

geoffromer commented on explorer/ast/member.cpp:
```diff
@@ -14,19 +14,34 @@ Member::Member(Nonnull<const Declaration*> declaration)
+  } else if (const auto* named_valued = member_.dyn_cast<const NamedValue*>()) {
+    return named_valued->name;
+  } else {
+    return "";
```
I think returning an empty string here is likely to cause bugs; at a minimum I'd like this to `CHECK` fail instead, but ideally I'd like to restructure the API so you can't write those bugs in the first place. Specifically, I think we should get rid of `name()` on both `Member` and `Component`, and replace it with two separate operations:

- `void Print(llvm::raw_ostream& out)`, which prints the `Member`/`Component`.
- `auto IsNamed(std::string_view name) -> bool`, which returns true if the `Member`/`Component` is named `name`.

I think those two are enough to support all the current use cases for `name`.

Pixep replied:
Thank you, I did add both methods, but had to keep `name()` for now, because `GetMember` (now `GetNamedMember`) relies heavily on having a string version of the name available for various types and situation, provided to multiple methods. So that wasn't actually practical due to that one method content.

geoffromer commented on explorer/ast/member.h:
```diff
@@ -33,16 +43,20 @@ class Member {
   // The name of the member.
   auto name() const -> std::string_view;
+  // The index of the member, if any
+  auto index() const -> std::optional<size_t>;
```
I think we should say that this method requires that `*this` represents a positional member rather than a named member, have it `CHECK`-fail internally if that's not the case, and otherwise return a `size_t`. That way the caller doesn't have to deal with unwrapping the `optional`.

Pixep replied:
Done. Provided a couple of getters: `hasIndex()` and `hasName()` to discriminate between in places where they are used.

Pixep committed:
Explorer: improve Member API for named vs positional members

Given Member may or not be named, provide a better API for its usage:
- Member::Print(out)
- Member::IsNamed(std::string_view name) -> bool
- Member::isPositional() -> bool

Update dependent code to rely less on 'name()' by default

Pixep committed:
Explorer: use better names for getters

And refactor a bit

Pixep committed:
Explorer: enforce having a named member for `MemberName`

Pixep committed:
Explorer: Component API cleanup

Pixep committed:
Explorer: fix comment typo

Pixep commented:
Note: https://github.com/carbon-language/carbon-lang/pull/2421 is an incremental change on this one that takes this work to the next level, as I personally don't feel like the current design would be sustainable. Kept separate to reduce the changeset.

geoffromer commented on explorer/interpreter/value.cpp:
```diff
@@ -192,13 +217,12 @@ static auto SetFieldImpl(
-      if (index < 0 || static_cast<size_t>(index) >= elements.size()) {
-        return ProgramError(source_loc) << "index " << (*path_begin).name()
-                                        << " out of range in " << *value;
+      CARBON_CHECK((*path_begin).member().hasPosition())
```
As above, this `CARBON_CHECK` seems redundant since `index()` is about to do the same thing.

Pixep replied:
Removed

geoffromer commented on explorer/interpreter/value.cpp:
```diff
@@ -33,12 +34,34 @@ auto StructValue::FindField(std::string_view name) const
+                                SourceLocation source_loc)
     -> ErrorOr<Nonnull<const Value*>> {
-  std::string_view f = field.name();
+  CARBON_CHECK(field.member().hasPosition()) << "Invalid non-positional member";
```
I'd prefer to keep precondition checks like this inside the methods of `Member` if possible, because that can enable us to omit accessors like `hasPosition`, and avoid having redundant checks spread across the call stack, which can get unwieldy. In this case, `index()` already has a `CARBON_CHECK` that handles this, so I think we can drop this one.

Checking these preconditions at every level of the callstack can get unwieldy, and forces us to expose accessors like `hasPosition` that might not otherwise be necessary. I'd prefer to drop this, and let it be handled by the `index()` call below.

Pixep replied:
Right, that wouldn't scale super well. Removed.

geoffromer commented on explorer/interpreter/value.cpp:
```diff
@@ -33,12 +34,34 @@ auto StructValue::FindField(std::string_view name) const
+                           SourceLocation source_loc,
+                           Nonnull<const Value*> me_value)
+    -> ErrorOr<Nonnull<const Value*>> {
+  CARBON_CHECK(field.member().hasName()) << "Invalid unnamed member";
```
As above, I think we can drop this, because the `name()` call on the next line will take care of it.

Pixep replied:
Removed

Pixep committed:
Explorer: remove superfluous CARBON_CHECKs

Pixep commented:
@geoffromer Thanks for the review! All 3 checks removed.

Pixep committed:
Merge branch 'trunk' into feature/tuple-member-variant

jonmeow commented on explorer/ast/member.h:
```diff
@@ -25,6 +26,16 @@ struct NamedValue {
+// field
+struct IndexedValue {
+  // The field index.
+  size_t index;
```
```suggestion
  int index;
```

I believe `int` is more consistent with style and surrounding code: https://google.github.io/styleguide/cppguide.html#Integer_Types

Pixep replied:
That's going to (somewhat) limit our capacity to address content beyond 2^31 for tuples, and needs `size_t` casts here and there. But consistency might be preferred. Change done, can always be reverted if needed.

jonmeow commented on explorer/interpreter/interpreter.cpp:
```diff
@@ -463,14 +463,10 @@ auto Interpreter::StepLvalue() -> ErrorOr<Success> {
-            object.SubobjectAddress(Member(tuple_field_as_struct_field));
+        const auto index = cast<IntValue>(*act.results()[1]).value();
+        auto* tuple_field = arena_->New<IndexedValue>(
+            IndexedValue{static_cast<size_t>(index), &exp.static_type()});
```
```suggestion
            IndexedValue{index, &exp.static_type()});
```

Note that if you switch index to `int` in general, this cast shouldn't be needed.

Pixep replied:
Done

jonmeow commented on explorer/ast/member.h:
```diff
@@ -25,6 +26,16 @@ struct NamedValue {
 };
 
+// A IndexedValue represents a value identified by an index, such as a tuple
+// field
```
```suggestion
// field.
```

Pixep replied:
Done

jonmeow commented on explorer/ast/member.h:
```diff
@@ -33,16 +44,30 @@ class Member {
+  // Prints the Member
+  void Print(llvm::raw_ostream& out) const;
+
+  // Return whether the member is positional, i.e. has an index.
+  auto hasPosition() const -> bool;
+  // Return whether the member is named, i.e. has a name.
+  auto hasName() const -> bool;
```
Names should either be accessor-named if trivial (has_name) or normal function named (HasName)

Pixep replied:
Git it. Done

jonmeow commented on explorer/ast/member.h:
```diff
@@ -33,16 +44,30 @@ class Member {
+  // The name of the member. Requires *this to represent a named member.
   auto name() const -> std::string_view;
+  // The index of the member. Requires *this to represent a positional member.
+  auto index() const -> size_t;
```
```suggestion
  auto index() const -> int;
```

Pixep replied:
Done

jonmeow reviewed (commented):
Coming over here from #2361. You can ignore my misgivings about IndexedValue there given geoffromer's review here, but some style comments remain.

Pixep committed:
Explorer: fix method naming & int type preference

Pixep commented:
> Coming over here from #2361. You can ignore my misgivings about IndexedValue there given geoffromer's review here, but some style comments remain.

Addressed comments.
The following https://github.com/carbon-language/carbon-lang/pull/2421 drops `IndexedValue`, so that might ease your concerns :).

geoffromer commented:
Can you update the PR description? "Adds `FieldPath::index() -> std::optional<size_t>`" looks out of date. Once that's done, I can merge.

Pixep commented:
> Can you update the PR description? "Adds `FieldPath::index() -> std::optional<size_t>`" looks out of date. Once that's done, I can merge.

Nice catch, Done
